clientconn_test.go

```diff
@@ -820,3 +820,135 @@ func TestBackoffCancel(t *testing.T) {
+		t.Fatalf("Error while listening. Err: %v", err)
+	}
+	defer lis.Close()
+	testFinished := make(chan struct{})
+	backoffCaseReady := make(chan struct{}, 1)
+	defer close(backoffCaseReady)
+	killFirstConnection := make(chan struct{})
+	killSecondConnection := make(chan struct{})
+
+	// Launch the server.
+	go func() {
+		// Establish a successful connection so that we enter READY. We need to get
+		// to READY so that we can get a client back for us to introspect later (as
+		// opposed to just CONNECTING).
+		conn, err := lis.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		defer conn.Close()
+
+		framer := http2.NewFramer(conn, conn)
+		if err := framer.WriteSettings(http2.Setting{}); err != nil {
+			t.Errorf("Error while writing settings frame. %v", err)
+			return
+		}
+
+		select {
+		case <-testFinished:
+			return
+		case <-killFirstConnection:
+		}
+
+		// Close the conn to cause onShutdown, causing us to enter TRANSIENT FAILURE. Note that we are not in
+		// WaitForHandshake at this point because the preface was sent successfully.
+		conn.Close()
+
+		// We have to re-accept and re-close the connection because the first re-connect after a successful handshake
+		// has no backoff. So, we need to get to the second re-connect after the successful handshake for our infinite
+		// backoff to happen.
+		conn, err = lis.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		err = conn.Close()
+		if err != nil {
+			t.Error(err)
+		}
+
+		// The client should now be headed towards backoff.
+		backoffCaseReady <- struct{}{}
+
+		// Re-connect (without server preface).
+		conn, err = lis.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+
+		// Close the conn to cause onShutdown, causing us to enter TRANSIENT FAILURE. Note that we are in
+		// WaitForHandshake at this point because the preface has not been sent yet.
+		select {
+		case <-testFinished:
+			return
+		case <-killSecondConnection:
+		}
+		conn.Close()
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	client, err := DialContext(ctx, lis.Addr().String(), WithInsecure(), WithWaitForHandshake(), WithBlock(), withBackoff(backoffForever{}))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	// It should start in READY because the server sends the server preface.
+	if got, want := client.GetState(), connectivity.Ready; got != want {
+		t.Fatalf("expected addrconn state to be %v, was %v", want, got)
+	}
+
+	// Once the connection is killed, it should go:
+	// READY -> CONNECTING -> TRANSIENT FAILURE (no backoff) -> CONNECTING -> TRANSIENT FAILURE (infinite backoff)
+	// The first TRANSIENT FAILURE is triggered by closing a channel. Then, we wait for the server to let us know
+	// when the client has progressed past the first failure (which does not get backoff, because handshake was
+	// successful).
+	close(killFirstConnection)
+	client.WaitForStateChange(ctx, connectivity.Ready)
```

Comment:
We should probably be checking the return value of all of these calls, unfortunately.  Maybe a helper function that you also pass `t` to to avoid making the code full of `if !WFSC { Fatal }`s?  It would be a good chance to try out `T.Helper()` (which I only learned about recently, at least).

Reply:
My thinking here was that the return value doesn't really tell us much. AFAICT, right after close(killFirstConnection), the connection could be in any part of READY -> TRANSIENT FAILURE -> CONNECTING -> TRANSIENT FAILURE.

I can't guarantee that it will be in READY when I begin waiting, I only want to make sure it's _not_  going to start asserting making the TRANSIENT FAILURE assertion until we know for sure it's the second one, not the first one.

Now that I write that out, I see a cleaner way to write this out. Instead of:

```
close(killFirstConnection)
client.WaitForStateChange(ctx, connectivity.Ready)
client.WaitForStateChange(ctx, connectivity.Connecting)
<-backoffCaseReady
client.WaitForStateChange(ctx, connectivity.Connecting)
if got, want := client.GetState(), connectivity.TransientFailure; got != want {
	t.Fatalf("expected addrconn state to be %v, was %v", want, got)
}
```

We can write:

```
close(killFirstConnection)
client.WaitForStateChange(ctx, connectivity.Ready)
client.WaitForStateChange(ctx, connectivity.TransientFailure)
client.WaitForStateChange(ctx, connectivity.Connecting)
if got, want := client.GetState(), connectivity.TransientFailure; got != want {
	t.Fatalf("expected addrconn state to be %v, was %v", want, got)
}
```

WDYT? I've tested it with `-cpu 1,4 -count 10000` and it seems ok.